File: python/tests/deduplication_test.py
"""
Tests for deduplication functionality.

Tests simple mode, throttle mode, and debounce mode deduplication.
"""

from bullmq import Queue
from uuid import uuid4

import asyncio
import redis.asyncio as redis
import unittest
import os

queueName = ""
prefix = os.environ.get('BULLMQ_TEST_PREFIX') or "bull"

class TestDeduplication(unittest.IsolatedAsyncioTestCase):

    def setUp(self):
        print("Setting up test queue")
        queueName = f"__test_queue__{uuid4().hex}"

    async def asyncTearDown(self):
        connection = redis.Redis(host='localhost')
        await connection.flushdb()

    async def test_simple_mode_deduplication(self):
        """Test simple mode where deduplication lasts until job is completed"""
        queue = Queue(queueName, {"prefix": prefix})
        
        # Add first job with deduplication
        job1 = await queue.add("test-job", {"color": "white"}, {
            "deduplication": {"id": "customValue"}
        })
        
        # Try to add second job with same deduplication ID
        job2 = await queue.add("test-job", {"color": "black"}, {
            "deduplication": {"id": "customValue"}
        })
        
        # The second job should be deduplicated (return the first job's ID)
        self.assertEqual(job1.id, job2.id)
        
        # Verify only one job exists in the queue
        jobs = await queue.getJobs(["waiting"])
        self.assertEqual(len(jobs), 1)
        self.assertEqual(jobs[0].id, job1.id)
        
        await queue.close()

    async def test_throttle_mode_deduplication(self):
        """Test throttle mode with TTL"""
        queue = Queue(queueName, {"prefix": prefix})
        
        # Add first job with deduplication and 1 second TTL
        job1 = await queue.add("test-job", {"color": "white"}, {
            "deduplication": {"id": "throttleTest", "ttl": 1000}
        })
        
        # Try to add second job immediately (should be deduplicated)
        job2 = await queue.add("test-job", {"color": "black"}, {
            "deduplication": {"id": "throttleTest", "ttl": 1000}
        })
        
        self.assertEqual(job1.id, job2.id)
        
        # Wait for TTL to expire (with small buffer)
        await asyncio.sleep(1.1)
        
        # Now we should be able to add a new job
        job3 = await queue.add("test-job", {"color": "red"}, {
            "deduplication": {"id": "throttleTest", "ttl": 1000}
        })
        
        self.assertNotEqual(job1.id, job3.id)
        
        await queue.close()

    async def test_debounce_mode_with_replace(self):
        """Test debounce mode where job data is replaced"""
        queue = Queue(queueName, {"prefix": prefix})
        
        # Add multiple jobs with replace option
        for i in range(1, 6):
            await queue.add(
                "test-job",
                {"color": f"white {i}"},
                {
                    "deduplication": {
                        "id": "debounceTest",
                        "ttl": 2000,
                        "extend": True,
                        "replace": True
                    },
                    "delay": 2000
                }
            )
        
        # Wait a bit to ensure all jobs are processed
        await asyncio.sleep(0.1)
        
        # Only one job should exist in delayed state
        delayed_jobs = await queue.getJobs(["delayed"])
        self.assertEqual(len(delayed_jobs), 1)
        
        # The job should have the latest data
        self.assertEqual(delayed_jobs[0].data["color"], "white 5")
        
        await queue.close()

    async def test_deduplication_with_extend(self):
        """Test deduplication with extend option"""
        queue = Queue(queueName, {"prefix": prefix})
        
        # Add first job with extend option
        job1 = await queue.add("test-job", {"value": 1}, {
            "deduplication": {"id": "extendTest", "ttl": 1000, "extend": True}
        })
        
        # Wait half the TTL
        await asyncio.sleep(0.6)
        
        # Add second job (should extend TTL)
        job2 = await queue.add("test-job", {"value": 2}, {
            "deduplication": {"id": "extendTest", "ttl": 1000, "extend": True}
        })
        
        # Should be deduplicated
        self.assertEqual(job1.id, job2.id)
        
        # Wait original TTL period (should still be deduplicated)
        await asyncio.sleep(0.6)
        
        # Try to add another job (should still be deduplicated due to extended TTL)
        job3 = await queue.add("test-job", {"value": 3}, {
            "deduplication": {"id": "extendTest", "ttl": 1000, "extend": True}
        })
        
        self.assertEqual(job1.id, job3.id)
        
        await queue.close()

    async def test_different_deduplication_ids(self):
        """Test that different deduplication IDs create different jobs"""
        queue = Queue(queueName, {"prefix": prefix})
        
        job1 = await queue.add("test-job", {"value": 1}, {
            "deduplication": {"id": "id1"}
        })
        
        job2 = await queue.add("test-job", {"value": 2}, {
            "deduplication": {"id": "id2"}
        })
        
        # Different IDs should create different jobs
        self.assertNotEqual(job1.id, job2.id)
        
        # Verify both jobs exist
        jobs = await queue.getJobs(["waiting"])
        self.assertEqual(len(jobs), 2)
        
        await queue.close()

    async def test_deduplication_without_ttl(self):
        """Test deduplication without TTL (lasts until job completion)"""
        queue = Queue(queueName, {"prefix": prefix})
        
        job1 = await queue.add("test-job", {"value": 1}, {
            "deduplication": {"id": "noTtlTest"}
        })
        
        # Wait a bit
        await asyncio.sleep(1)
        
        # Should still be deduplicated
        job2 = await queue.add("test-job", {"value": 2}, {
            "deduplication": {"id": "noTtlTest"}
        })
        
        self.assertEqual(job1.id, job2.id)
        
        await queue.close()

    async def test_returns_none_when_no_deduplicated_job(self):
        result = await self.queue.getDeduplicationJobId("missing-id")
        self.assertIsNone(result)

    async def test_returns_job_id_for_simple_mode(self):
        job = await self.queue.add(
            "test-job", {}, {"deduplication": {"id": "my-dedup-id"}}
        )

        result = await self.queue.getDeduplicationJobId("my-dedup-id")

        self.assertEqual(result, job.id)

    async def test_returns_job_id_for_throttle_mode_within_ttl(self):
        job = await self.queue.add(
            "test-job", {}, {"deduplication": {"id": "throttle-id", "ttl": 5000}}
        )

        result = await self.queue.getDeduplicationJobId("throttle-id")

        self.assertEqual(result, job.id)

    async def test_returns_none_after_job_completes_simple_mode(self):
        await self.queue.add(
            "test-job", {}, {"deduplication": {"id": "completing-id"}}
        )
        worker = Worker(queueName, None, {"prefix": prefix})
        token = 'my-token'

        job = await worker.getNextJob(token)

        await job.moveToCompleted("done", token="test-token", fetchNext=False)
        # or drive it through a real Worker if that's the repo's convention
        # for exercising completion in this test file

        result = await self.queue.getDeduplicationJobId("completing-id")

        self.assertIsNone(result)

    async def test_deprecated_get_debounce_job_id_delegates(self):
        job = await self.queue.add(
            "test-job", {}, {"deduplication": {"id": "alias-id"}}
        )

        result = await self.queue.getDebounceJobId("alias-id")

        self.assertEqual(result, job.id)

    async def test_ignores_deduplication_id_from_other_queue(self):
        other_queue_name = f"{self.queueName}-other"
        other_queue = Queue(other_queue_name, {"connection": redis_connection})
        try:
            await other_queue.add(
                "test-job", {}, {"deduplication": {"id": "shared-id"}}
            )
            result = await self.queue.getDeduplicationJobId("shared-id")
            self.assertIsNone(result)
        finally:
            await other_queue.close()

if __name__ == '__main__':
    unittest.main()

File: rust/tests/deduplication_test.rs
//! Deduplication tests.

mod common;

use std::sync::Arc;
use std::time::Duration;

use bullmq::worker::{CancellationToken, ProcessorFn};
use bullmq::{DeduplicationOptions, Job, JobOptions, Queue, QueueOptions, Worker, WorkerOptions};
use common::{cleanup_queue, test_connection, test_queue_name};
use tokio::sync::mpsc;

// ═══════════════════════════════════════════════════════════════════════════
// Basic deduplication
// ═══════════════════════════════════════════════════════════════════════════

#[tokio::test]
async fn test_dedup_prevents_duplicate_jobs() {
    let name = test_queue_name();
    let conn_opts = test_connection();
    let queue_opts = QueueOptions {
        connection: conn_opts.clone(),
        ..Default::default()
    };
    let queue = Queue::new(&name, queue_opts).await.unwrap();

    // Add two jobs with the same deduplication ID
    let opts1 = JobOptions {
        deduplication: Some(DeduplicationOptions {
            id: "dedup-1".to_string(),
            ..Default::default()
        }),
        ..Default::default()
    };
    let opts2 = JobOptions {
        deduplication: Some(DeduplicationOptions {
            id: "dedup-1".to_string(),
            ..Default::default()
        }),
        ..Default::default()
    };

    let id1 = queue
        .add("job1", serde_json::json!({"first": true}), Some(opts1))
        .await
        .unwrap()
        .id()
        .to_string();
    let id2 = queue
        .add("job2", serde_json::json!({"second": true}), Some(opts2))
        .await
        .unwrap()
        .id()
        .to_string();

    // Both should return the same job ID (second was deduplicated)
    assert_eq!(id1, id2);

    // Only 1 job should be in the queue
    let counts = queue.get_job_counts().await.unwrap();
    assert_eq!(counts.waiting, 1);

    cleanup_queue(&queue).await;
}

#[tokio::test]
async fn test_dedup_different_ids_create_separate_jobs() {
    let name = test_queue_name();
    let conn_opts = test_connection();
    let queue_opts = QueueOptions {
        connection: conn_opts.clone(),
        ..Default::default()
    };
    let queue = Queue::new(&name, queue_opts).await.unwrap();

    let opts1 = JobOptions {
        deduplication: Some(DeduplicationOptions {
            id: "dedup-a".to_string(),
            ..Default::default()
        }),
        ..Default::default()
    };
    let opts2 = JobOptions {
        deduplication: Some(DeduplicationOptions {
            id: "dedup-b".to_string(),
            ..Default::default()
        }),
        ..Default::default()
    };

    let id1 = queue
        .add("job1", serde_json::json!({"a": 1}), Some(opts1))
        .await
        .unwrap()
        .id()
        .to_string();
    let id2 = queue
        .add("job2", serde_json::json!({"b": 2}), Some(opts2))
        .await
        .unwrap()
        .id()
        .to_string();

    // Different dedup IDs → different jobs
    assert_ne!(id1, id2);

    let counts = queue.get_job_counts().await.unwrap();
    assert_eq!(counts.waiting, 2);

    cleanup_queue(&queue).await;
}

// ═══════════════════════════════════════════════════════════════════════════
// Deduplication with TTL
// ═══════════════════════════════════════════════════════════════════════════

#[tokio::test]
async fn test_dedup_with_ttl_allows_after_expiry() {
    let name = test_queue_name();
    let conn_opts = test_connection();
    let queue_opts = QueueOptions {
        connection: conn_opts.clone(),
        ..Default::default()
    };
    let queue = Queue::new(&name, queue_opts).await.unwrap();

    // First job with 500ms TTL
    let opts1 = JobOptions {
        deduplication: Some(DeduplicationOptions {
            id: "ttl-dedup".to_string(),
            ttl: Some(500),
            ..Default::default()
        }),
        ..Default::default()
    };
    queue
        .add("job1", serde_json::json!({"first": true}), Some(opts1))
        .await
        .unwrap();

    // Immediately add a duplicate → should be deduplicated
    let opts2 = JobOptions {
        deduplication: Some(DeduplicationOptions {
            id: "ttl-dedup".to_string(),
            ttl: Some(500),
            ..Default::default()
        }),
        ..Default::default()
    };
    queue
        .add("job2", serde_json::json!({"second": true}), Some(opts2))
        .await
        .unwrap();

    let counts = queue.get_job_counts().await.unwrap();
    assert_eq!(counts.waiting, 1, "Duplicate should be prevented");

    // Wait for TTL to expire
    tokio::time::sleep(Duration::from_millis(600)).await;

    // Now add again → should create a new job
    let opts3 = JobOptions {
        deduplication: Some(DeduplicationOptions {
            id: "ttl-dedup".to_string(),
            ttl: Some(500),
            ..Default::default()
        }),
        ..Default::default()
    };
    queue
        .add("job3", serde_json::json!({"third": true}), Some(opts3))
        .await
        .unwrap();

    let counts = queue.get_job_counts().await.unwrap();
    assert_eq!(
        counts.waiting, 2,
        "After TTL expiry, new job should be created"
    );

    cleanup_queue(&queue).await;
}

// ═══════════════════════════════════════════════════════════════════════════
// Deduplication with replace
// ═══════════════════════════════════════════════════════════════════════════

#[tokio::test]
async fn test_dedup_replace_updates_delayed_job() {
    let name = test_queue_name();
    let conn_opts = test_connection();
    let queue_opts = QueueOptions {
        connection: conn_opts.clone(),
        ..Default::default()
    };
    let queue = Queue::new(&name, queue_opts).await.unwrap();

    // First job delayed with dedup
    let opts1 = JobOptions {
        delay: Some(60_000),
        deduplication: Some(DeduplicationOptions {
            id: "replace-dedup".to_string(),
            ..Default::default()
        }),
        ..Default::default()
    };
    let id1 = queue
        .add("job1", serde_json::json!({"v": 1}), Some(opts1))
        .await
        .unwrap()
        .id()
        .to_string();

    // Replace with a new delayed job
    let opts2 = JobOptions {
        delay: Some(60_000),
        deduplication: Some(DeduplicationOptions {
            id: "replace-dedup".to_string(),
            replace: Some(true),
            ..Default::default()
        }),
        ..Default::default()
    };
    let id2 = queue
        .add("job2", serde_json::json!({"v": 2}), Some(opts2))
        .await
        .unwrap()
        .id()
        .to_string();

    // New job should have a different ID
    assert_ne!(id1, id2);

    // Only 1 delayed job should exist (the replacement)
    let counts = queue.get_job_counts().await.unwrap();
    assert_eq!(counts.delayed, 1);

    // Verify the existing job has the updated data
    let job = queue.get_job(&id2).await.unwrap().unwrap();
    assert_eq!(job.data()["v"], 2);

    cleanup_queue(&queue).await;
}

// ═══════════════════════════════════════════════════════════════════════════
// Remove deduplication key
// ═══════════════════════════════════════════════════════════════════════════

#[tokio::test]
async fn test_remove_deduplication_key() {
    let name = test_queue_name();
    let conn_opts = test_connection();
    let queue_opts = QueueOptions {
        connection: conn_opts.clone(),
        ..Default::default()
    };
    let queue = Queue::new(&name, queue_opts).await.unwrap();

    // Add a job with dedup (no TTL → key persists indefinitely)
    let opts = JobOptions {
        deduplication: Some(DeduplicationOptions {
            id: "persistent-dedup".to_string(),
            ..Default::default()
        }),
        ..Default::default()
    };
    let job_id = queue
        .add("job1", serde_json::json!({"v": 1}), Some(opts))
        .await
        .unwrap()
        .id()
        .to_string();

    // Verify dedup is active — adding same dedup ID returns same job
    let opts2 = JobOptions {
        deduplication: Some(DeduplicationOptions {
            id: "persistent-dedup".to_string(),
            ..Default::default()
        }),
        ..Default::default()
    };
    let id2 = queue
        .add("job2", serde_json::json!({"v": 2}), Some(opts2))
        .await
        .unwrap()
        .id()
        .to_string();
    assert_eq!(job_id, id2);

    // Remove the dedup key
    let removed = queue
        .remove_deduplication_key("persistent-dedup", &job_id)
        .await
        .unwrap();
    assert!(removed);

    // Now a new job should be created
    let opts3 = JobOptions {
        deduplication: Some(DeduplicationOptions {
            id: "persistent-dedup".to_string(),
            ..Default::default()
        }),
        ..Default::default()
    };
    let id3 = queue
        .add("job3", serde_json::json!({"v": 3}), Some(opts3))
        .await
        .unwrap()
        .id()
        .to_string();
    assert_ne!(job_id, id3);

    cleanup_queue(&queue).await;
}

// ═══════════════════════════════════════════════════════════════════════════
// Deduplication with worker processing
// ═══════════════════════════════════════════════════════════════════════════

#[tokio::test]
async fn test_dedup_key_removed_after_job_completes() {
    let name = test_queue_name();
    let conn_opts = test_connection();
    let queue_opts = QueueOptions {
        connection: conn_opts.clone(),
        ..Default::default()
    };
    let queue = Queue::new(&name, queue_opts).await.unwrap();

    // Add job with dedup (no TTL → cleaned up on completion by Lua moveToFinished)
    let opts = JobOptions {
        deduplication: Some(DeduplicationOptions {
            id: "complete-dedup".to_string(),
            ..Default::default()
        }),
        ..Default::default()
    };
    queue
        .add("job1", serde_json::json!({"v": 1}), Some(opts))
        .await
        .unwrap();

    // Process it
    let (tx, mut rx) = mpsc::channel(5);
    let processor: ProcessorFn = Arc::new(move |_job: Job, _token: CancellationToken| {
        let tx = tx.clone();
        Box::pin(async move {
            tx.send(()).await.unwrap();
            Ok(serde_json::Value::Null)
        })
    });

    let worker_opts = WorkerOptions {
        connection: conn_opts.clone(),
        autorun: true,
        ..Default::default()
    };
    let worker = Worker::new(&name, processor, worker_opts).await.unwrap();

    tokio::time::timeout(Duration::from_secs(3), rx.recv())
        .await
        .unwrap();
    tokio::time::sleep(Duration::from_millis(100)).await;

    // After completion, dedup key should be removed → new job can be added
    let opts2 = JobOptions {
        deduplication: Some(DeduplicationOptions {
            id: "complete-dedup".to_string(),
            ..Default::default()
        }),
        ..Default::default()
    };
    let _id2 = queue
        .add("job2", serde_json::json!({"v": 2}), Some(opts2))
        .await
        .unwrap();

    // Should be a new job (not deduplicated)
    let counts = queue.get_job_counts().await.unwrap();
    // Either the new job is waiting or already picked up by the worker
    assert!(
        counts.waiting + counts.active >= 1,
        "New job should have been created after dedup key cleared"
    );

    worker.close(5000).await.unwrap();
    cleanup_queue(&queue).await;
}

// ═══════════════════════════════════════════════════════════════════════════
// Deprecated debounce aliases
// ═══════════════════════════════════════════════════════════════════════════

#[tokio::test]
async fn test_get_debounce_job_id_alias() {
    let name = test_queue_name();
    let queue = Queue::new(
        &name,
        QueueOptions {
            connection: test_connection(),
            ..Default::default()
        },
    )
    .await
    .unwrap();

    assert_eq!(queue.get_debounce_job_id("missing").await.unwrap(), None);

    let job_id = queue
        .add(
            "job1",
            serde_json::json!({"v": 1}),
            Some(JobOptions {
                deduplication: Some(DeduplicationOptions {
                    id: "debounce-1".to_string(),
                    ..Default::default()
                }),
                ..Default::default()
            }),
        )
        .await
        .unwrap()
        .id()
        .to_string();

    // The deprecated alias returns the same value as get_deduplication_job_id.
    assert_eq!(
        queue.get_debounce_job_id("debounce-1").await.unwrap(),
        Some(job_id.clone())
    );
    assert_eq!(
        queue.get_deduplication_job_id("debounce-1").await.unwrap(),
        Some(job_id)
    );

    cleanup_queue(&queue).await;
}

#[tokio::test]
async fn test_remove_debounce_key_alias() {
    let name = test_queue_name();
    let queue = Queue::new(
        &name,
        QueueOptions {
            connection: test_connection(),
            ..Default::default()
        },
    )
    .await
    .unwrap();

    queue
        .add(
            "job1",
            serde_json::json!({"v": 1}),
            Some(JobOptions {
                deduplication: Some(DeduplicationOptions {
                    id: "debounce-rm".to_string(),
                    ..Default::default()
                }),
                ..Default::default()
            }),
        )
        .await
        .unwrap();

    // Plain DEL — returns 1 when the key existed, 0 afterwards.
    assert_eq!(queue.remove_debounce_key("debounce-rm").await.unwrap(), 1);
    assert_eq!(queue.remove_debounce_key("debounce-rm").await.unwrap(), 0);
    assert_eq!(
        queue.get_debounce_job_id("debounce-rm").await.unwrap(),
        None
    );

    cleanup_queue(&queue).await;
}
